src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt

```diff
@@ -167,20 +170,106 @@ class ProjectService(
 
-        // TODO (see #314): correct status handling: if a project is archived, you should only be able to activate it
-        // and if it is active, you can do everything, but deleting / restoring it never allowed.
-        if (project.status == ProjectStatus.PROJECT_STATUS_DELETED) {
-            throw FailedPreconditionException("The project with the id ${request.project.id} is deleted.")
+        val project = repo.getProjectById(projectId).getOrThrow()
+        val currentStatus = project.status
+
+        val fieldMask = request.mask.pathsList
+        val requestedStatus = request.project.status
+
+        validateProjectUpdate(currentStatus, requestedStatus, fieldMask)
+
+        val finalStatus = determineFinalStatus(projectId, requestedStatus)
+        val finalRequest = request.copy {
+            this.project = request.project.copy {
+                this.status = finalStatus
+            }
         }
-        if (request.project.status == ProjectStatus.PROJECT_STATUS_DELETED) {
-            throw FailedPreconditionException("The project status can not be set to deleted by an update call.")
+
+        repo.updateProject(finalRequest).toGrpcProject()
+    }
+
+    /**
+     * Validates the update of a project based on the current status and the requested status.
+     *
+     * If the project is `PROJECT_STATUS_DELETED`, nothing can be updated.
+     * If the project is `PROJECT_STATUS_ARCHIVED`, only the status can be updated (back to an active project).
+     * If the project is `PROJECT_STATUS_ACTIVE`, then everything can be updated.
+     * If the project is `PROJECT_STATUS_ACTIVE_LOCKED`, then everything except the slr settings can be updated.
+     *
+     * In addition, the status can never be set to `PROJECT_STATUS_DELETED` via the update method.
+     *
+     * @throws FailedPreconditionException if the update fails for any reason.
+     */
+    private fun validateProjectUpdate(
+        currentStatus: ProjectStatus,
+        requestedStatus: ProjectStatus,
+        fieldMask: List<String>,
+    ) {
+        require(!(fieldMask.contains("project.status") && requestedStatus == ProjectStatus.PROJECT_STATUS_DELETED)) {
+            "The project status cannot be set to DELETED via the update method. Use SoftDeleteProject instead."
         }
 
-        repo.updateProject(request, project.status).toGrpcProject()
+        when (currentStatus) {
+            ProjectStatus.PROJECT_STATUS_DELETED -> {
+                throw FailedPreconditionException("The update failed because the project has been deleted.")
+            }
+
+            ProjectStatus.PROJECT_STATUS_ARCHIVED -> {
+                val isOnlyStatusUpdate = fieldMask.size == 1 && fieldMask.contains("project.status")
+                if (!isOnlyStatusUpdate) {
+                    throw FailedPreconditionException(
+                        "For an archived project, only the 'status' field can be updated.",
```

Comment:
```suggestion
                        "The project is archived and therefore only the 'status' field can be updated.",
```
What do you think of all these suggestions? Better or worse?

Reply:
Better